File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {

        if(s.length() == 0 || s == null){
            return 0;
        }

        Map<Character,Integer> charFreqMap = new HashMap<>();

        int windowLeft = 0;
        int longestSubString = 0;

        int n = s.length();

        for(int windowRight = 0;windowRight < n;windowRight++){
            char ch = s.charAt(windowRight);

            charFreqMap.put(ch , charFreqMap.getOrDefault(ch , 0) + 1);

            while(charFreqMap.get(ch) > 1){
                char leftChar = s.charAt(windowLeft);
                charFreqMap.put(leftChar , charFreqMap.get(leftChar) -1);
                windowLeft++;
            }

            longestSubString = Math.max(longestSubString , windowRight - windowLeft +1);

        }

        return longestSubString;

        


        
    }
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        max_len = 0
        left = right = 0
        seen = set()
        while right < len(s):
            while s[right] in seen:
                seen.remove(s[left])
                left += 1
            max_len = max(max_len, right - left + 1)
            seen.add(s[right])
            right += 1

        
        return max_len